extension/observer/endpointswatcher.go

```diff
@@ -15,89 +15,153 @@
-	latestEndpoints := ew.Endpointslister.ListEndpoints()
+// notifyOfEndpoints alerts subscribed Notify instances by their NotifyID of Endpoint events,
+// updating their internal store of last Endpoints advertised.
+func (ew *EndpointsWatcher) notifyOfEndpoints(notifyIDs ...NotifyID) {
+	latestEndpoints := ew.EndpointsLister.ListEndpoints()
+
+	for _, notifyID := range notifyIDs {
+		var notify Notify
+		if n, ok := ew.toNotify.Load(notifyID); !ok {
+			// an Unsubscribe() must have occurred during this call
+			continue
+		} else if notify, ok = n.(Notify); !ok {
+			ew.logger.Warn("failed to obtain notify instance from EndpointsWatcher", zap.Any("notify", n))
+			continue
+		}
 
-	// Create map from ID to endpoint for lookup.
-	latestEndpointsMap := make(map[EndpointID]bool, len(latestEndpoints))
-	for _, e := range latestEndpoints {
-		latestEndpointsMap[e.ID] = true
-	}
+		// Create map from ID to endpoint for lookup.
+		latestEndpointsMap := make(map[EndpointID]bool, len(latestEndpoints))
+		for _, e := range latestEndpoints {
+			latestEndpointsMap[e.ID] = true
+		}
 
-	var removedEndpoints, addedEndpoints, updatedEndpoints []Endpoint
-	// Iterate over the latest endpoints obtained. An endpoint needs
-	// to be added in case it is not already available in existingEndpoints.
-	for _, e := range latestEndpoints {
-		if existingEndpoint, ok := ew.existingEndpoints[e.ID]; !ok {
-			ew.existingEndpoints[e.ID] = e
-			addedEndpoints = append(addedEndpoints, e)
-		} else if !reflect.DeepEqual(existingEndpoint, e) {
-			// Collect updated endpoints.
-			ew.existingEndpoints[e.ID] = e
-			updatedEndpoints = append(updatedEndpoints, e)
+		le, _ := ew.existingEndpoints.LoadOrStore(notifyID, map[EndpointID]Endpoint{})
+		var storedEndpoints map[EndpointID]Endpoint
+		var ok bool
+		if storedEndpoints, ok = le.(map[EndpointID]Endpoint); !ok {
+			ew.logger.Warn("failed to load Endpoint store from EndpointsWatcher", zap.Any("endpoints", le))
+			continue
+		}
+		// copy to not modify sync.Map value directly (will be reloaded)
+		existingEndpoints := map[EndpointID]Endpoint{}
+		for id, endpoint := range storedEndpoints {
+			existingEndpoints[id] = endpoint
 		}
-	}
 
-	// If endpoint present in existingEndpoints does not exist in the latest
-	// list, it needs to be removed.
-	for id, e := range ew.existingEndpoints {
-		if !latestEndpointsMap[e.ID] {
-			delete(ew.existingEndpoints, id)
-			removedEndpoints = append(removedEndpoints, e)
+		var removedEndpoints, addedEndpoints, changedEndpoints []Endpoint
+		// Iterate over the latest endpoints obtained. An endpoint needs
+		// to be added or updated in case it is not already available in existingEndpoints or doesn't match
+		// the latest value.
+		for _, e := range latestEndpoints {
+			if existingEndpoint, ok := existingEndpoints[e.ID]; !ok {
+				existingEndpoints[e.ID] = e
+				addedEndpoints = append(addedEndpoints, e)
+			} else if !reflect.DeepEqual(existingEndpoint, e) {
```

Comment:
Ideally, `reflect` should be avoided where possible since there is some performance impact that come with it.
Is it possible to rewrite this so we don't need to use `reflect`?

I suspect this may also go away once `go1.18` is being adopted.

Reply:
Because `EndpointEnv` is a `map[string]interface{}` that's (at this time not arbitrarily) nested, it's convenient to use reflection and a pass without it would be pretty brittle as endpoints grow more complex and may look something like https://gist.github.com/rmfitzpatrick/f7dc44809361826f5f65bffb893cfb0a (not heavily tested). One thought is to take as many easy checks as possible and then use reflection when comparing the `EndpointEnv`?

Reply:
Committed the gist but I'd need to add more tests. I'm not immediately clear how `comparable` would help and am generally ok with the existing use of reflection considering this isn't really a hot path.